### ISSUE TITLE
Metrics names get modified for multi-output models

### ISSUE THREAD

**Issue by jbboin** [tensorflow/tensorflow#44241]
**System information**
- Have I written custom code: yes
- OS Platform and Distribution: Linux Ubuntu 16.04.5 LTS
- TensorFlow installed from: binary, installed with pip
- TensorFlow version (use command below): 2.3.0, 2.3.1 and nightly (2.4.0-dev20201022)
- Python version: 3.6.9
- CUDA/cuDNN version: 10.1
- GPU model and memory: NVIDIA T4, 16 GB

**Describe the current behavior**
When training a multi-output model with a metric in `tf.keras.metric`, the name of the metric is modified during training (the name of the output is added as a prefix to the name of the metric). This causes the model to get saved incorrectly when using `save()` (saving to a `h5` file). When the saved model is reloaded, the name of the output is added another time to the metric. So every time we save/load the metric name gets longer. In TF 2.3.0 and 2.3.1, this problem seems to only affect metrics in `tf.keras.metric` and metrics defined by a string (e.g. `accuracy`) don't seem to be affected. However in the nightly this problem seems to have got worse and affects both types of metrics.

**Describe the expected behavior**
If a multi-output model is saved to a `h5` file and reloaded, the metric names should be the same as in the original model.

**Standalone code to reproduce the issue**
```
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Input, Dense, Softmax
from tensorflow.keras.models import Model, load_model

# Create multi-output model
inp = Input(4)
m = Model(inp, [Softmax(name='head_0')(Dense(3)(inp)), Softmax(name='head_1')(Dense(5)(inp))])
metric = tf.keras.metrics.BinaryAccuracy()
m.compile(loss='mse', metrics = {'head_0': [metric, 'accuracy']})

print(f'Metric name: {metric.name}')

# Run one iteration
x = np.random.rand(2,4)
y = {'head_0': np.random.randint(2, size=(2,3)), 'head_1': np.random.randint(2, size=(2,5))}
m.fit(x, y, verbose=0)

print(f'Metric name: {metric.name}')

print(m.metrics_names)

# Save the model to a h5 file and reload it
model_fpath = 'model.h5'
m.save(model_fpath, save_format='h5')
m2 = load_model(model_fpath)

print(m2.metrics_names)
```

In this case we can see that `metric.name` gets modified after `fit` is called. The `metrics_names` before saving are initially:
```
['loss', 'head_0_loss', 'head_1_loss', 'head_0_binary_accuracy', 'head_0_accuracy']
```
After reloading in TF 2.3.x:
```
['loss', 'head_0_loss', 'head_1_loss', 'head_0_head_0_binary_accuracy', 'head_0_accuracy']
```
And after reloading in TF 2.4.0:
```
['loss', 'head_0_loss', 'head_1_loss', 'head_0_head_0_binary_accuracy', 'head_0_head_0_accuracy']
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/37d03cd116a93751dd730f4d647ced25/44241.ipynb). Thanks!

**Comment by nrothGIT** [comment]
Wanted to rebump! Any update here? Thanks so much for all your work on tflow, also :)

**Comment by oXwvdrbbj8S4wo9k8lSN** [comment]
Not a solution but at least a way to deal with the results:
```
def declutter_metric_name(metric_name, output_names):
    num_substitutions = 0
    for output_name in output_names:
        metric_name, number_of_subs_made = re.subn('({output_name}_){{2,}}'.format(output_name=output_name), output_name + '_', metric_name, count=1)
        num_substitutions += (number_of_subs_made>0)
    if num_substitutions>1:
        warnings.warn("Multiple substitutions detected! The result might not be valid.")
    return metric_name

def declutter_history(history, output_names):
    return {declutter_metric_name(metric_name, output_names): metric_values for metric_name, metric_values in history.items()}
```
Using these functions, multiple occurrences of output names in the metrics names can be corrected:
```
cluttered_metric_name = 'bacon_spam_spam_spam_spam_spam_spam_egg_and_spam'
print(declutter_metric_name(cluttered_metric_name, ["spam"]))
>>> 'bacon_spam_egg_and_spam'
```

**Comment by nrothGIT** [comment]
Hey! Thanks for writing! Yeah, this makes sense, but in theory you might want to have a name like 'bacon_spam_spam' and it becomes hard to disambiguate in general.

**Comment by maveyx** [comment]
I confirm this issue which popped up when I tried to re-compile the model after calling fit(), to make layer weights trainable/not trainable.
I use TF 2.2.
Any updates on that are very much appreciated. Thank you!

**Comment by jbboin** [comment]
As anecdotal evidence, we noticed this issue when upgrading from TF 2.1 to 2.3, so given your data point it seems that the issue was introduced by TF 2.2

**Comment by maveyx** [comment]
Ah sorry, I just checked my tensorflow version, it is TF 2.3!

**Comment by jbboin** [comment]
Oh ok, never mind then. In any case the issue wasn't there in TF 2.1.

**Comment by gouyelliot** [comment]
I found a workaround for the issue (tested in TS 2.4.0) by using the classes in `tensorflow.keras.metrics` and the `name` parameter:

```python
TENSORFLOW_METRICS = [
        tensorflow.keras.metrics.Accuracy(name='accuracy'),
        tensorflow.keras.metrics.Precision(name='precision'),
        tensorflow.keras.metrics.Recall(name='recall'),
        tensorflow.keras.metrics.AUC(curve='PR', name='auc_pr')
]

...

def train():
    modelA.compile(metrics=TENSORFLOW_METRICS, ...)
    modelB.compile(metrics=TENSORFLOW_METRICS, ...)
```

The data are shown on the same graph on TensorBoard too. Hope this helps !

**Comment by dellis23** [comment]
Thanks for reporting this.  It should be fixed with b8d17793a7aae010754d7776612af5b9b1f5252f and 1bd0748a42e2731e04b0b4d4b6fedcc34222492a.  Both should be available in the upcoming 2.5 release.
